packages/vaex-core/vaex/dataframe_protocol.py

```diff
@@ -0,0 +1,688 @@
+"""
+import enum
+import collections.abc
+import ctypes
+from typing import Any, Optional, Tuple, Dict, Iterable, Sequence
+
+import vaex
+import pyarrow
+import pyarrow as pa
+import numpy as np
+import pandas as pd
+
+
+# A typing protocol could be added later to let Mypy validate code using
+# `from_dataframe` better.
+DataFrameObject = Any
+ColumnObject = Any
+
+
+def from_dataframe_to_vaex(df: DataFrameObject, allow_copy: bool = True) -> vaex.dataframe.DataFrame:
+    """
+    Construct a vaex DataFrame from ``df`` if it supports ``__dataframe__``
+    """
+    if isinstance(df, vaex.dataframe.DataFrame):
+        return df
+
+    if not hasattr(df, "__dataframe__"):
+        raise ValueError("`df` does not support __dataframe__")
+
+    return _from_dataframe_to_vaex(df.__dataframe__(allow_copy=allow_copy))
+
+
+def _from_dataframe_to_vaex(df: DataFrameObject) -> vaex.dataframe.DataFrame:
+    """
+    Note: we need to implement/test support for bit/byte masks, chunk handling, etc.
+    """
+    # Iterate through the chunks
+    dataframe = []
+    _buffers = []
+    for chunk in df.get_chunks():
+
+        # We need a dict of columns here, with each column being an expression.
+        columns = dict()
+        _k = _DtypeKind
+        _buffers_chunks = []  # hold on to buffers, keeps memory alive
+        for name in chunk.column_names():
+            if not isinstance(name, str):
+                raise ValueError(f"Column {name} is not a string")
+            if name in columns:
+                raise ValueError(f"Column {name} is not unique")
+
+            col = chunk.get_column_by_name(name)
+            if col.dtype[0] in (_k.INT, _k.UINT, _k.FLOAT, _k.BOOL):
+                # Simple numerical or bool dtype, turn into arrow array
+                columns[name], _buf = convert_column_to_ndarray(col)
+            elif col.dtype[0] == _k.CATEGORICAL:
+                columns[name], _buf = convert_categorical_column(col)
+            # TODO
+            # elif col.dtype[0] == _k.STRING:
+            #    columns[name], _buf = convert_string_column(col)
+            else:
+                raise NotImplementedError(f"Data type {col.dtype[0]} not handled yet")
+
+            _buffers_chunks.append(_buf)
+
+        dataframe.append(vaex.from_dict(columns))
+        # chunk buffers are added to list of all buffers
+        _buffers.append(_buffers_chunks)
+
+    if df.num_chunks() == 1:
+        _buffers = _buffers[0]
+
+    df_new = vaex.concat(dataframe)
+    df_new._buffers = _buffers
+    return df_new
+
+
+class _DtypeKind(enum.IntEnum):
+    INT = 0
+    UINT = 1
+    FLOAT = 2
+    BOOL = 20
+    STRING = 21  # UTF-8
+    DATETIME = 22
+    CATEGORICAL = 23
+
+
+def convert_column_to_ndarray(col: ColumnObject) -> pa.Array:
+    """
+    Convert an int, uint, float or bool column to an arrow array
+    """
+    if col.offset != 0:
+        raise NotImplementedError("column.offset > 0 not handled yet")
```

Comment:
Before I force pushed, you asked if we used offset. I think we need them for string arrays, and for slicing bitmask arrays, see https://github.com/data-apis/dataframe-api/issues/67